File: quant1x/id/id64/id_test.py
# -*- coding: utf-8 -*-
# Copyright (c) Quant1X <wangfengxy@sina.cn>.
# Licensed under the MIT License.

from __future__ import annotations

import os
import tempfile
import threading
import unittest

from .generator import Generator
from .hlc import HLC, PAYLOAD_BITS
from .id import ID
from .option import with_clock, with_node_count, with_seq_seed, with_state_file, with_state_strict
from .state_store import PersistentState


class Id64Tests(unittest.TestCase):
    def test_hlc_rollback_monotonic(self) -> None:
        current = {"value": 1000}
        hlc = HLC(with_clock(lambda: current["value"]), with_seq_seed(9))

        prev_physical, prev_seq = hlc.now()
        current["value"] = 500  # 时钟回拨
        cur_physical, cur_seq = hlc.now()

        self.assertTrue(
            cur_physical > prev_physical
            or (cur_physical == prev_physical and cur_seq > prev_seq)
        )

    def test_uses_options_at_construction(self) -> None:
        current = {"value": 4321}
        hlc = HLC(with_clock(lambda: current["value"]), with_seq_seed(9))

        self.assertEqual(hlc.timestamp(), 4321)
        self.assertEqual(hlc.seq, 9)

    def test_node_count_derivation(self) -> None:
        cases = [
            (1024, 11, 11),
            (5000, 13, 9),
            (3, 2, 20),
            (131072, 18, 4),
        ]
        for count, worker_bits, seq_bits in cases:
            hlc = HLC(with_node_count(count))
            self.assertEqual(hlc.seq_bits, seq_bits, f"count={count}")
            gen = Generator(0, hlc)
            self.assertEqual(gen.worker_bits, worker_bits, f"count={count}")

    def test_node_count_too_large(self) -> None:
        with self.assertRaises(ValueError):
            HLC(with_node_count(262144))  # seqBits = 3 < 4

    def test_id_field_decoding(self) -> None:
        elapsed = 0x123456789A
        worker_bits = 11
        seq_bits = 11
        node_id = 0x1F
        seq = 0x2A

        value = (elapsed << PAYLOAD_BITS) | (node_id << seq_bits) | seq
        identifier = ID.from_int(value)

        self.assertEqual(identifier.physical(), elapsed)
        self.assertEqual(identifier.node_id(worker_bits), node_id)
        self.assertEqual(identifier.seq(worker_bits), seq)
        self.assertEqual(identifier.to_int(), value)

    def test_persistent_state_across_restart(self) -> None:
        current = {"value": 1000}
        with tempfile.TemporaryDirectory() as tmp:
            path = os.path.join(tmp, "state.bin")
            hlc1 = HLC(with_clock(lambda: current["value"]), with_state_file(path), with_seq_seed(9))
            p1, s1 = hlc1.now()
            hlc1.close()  # 快速路径为批量缓冲：优雅退出前刷盘

            hlc2 = HLC(with_clock(lambda: current["value"]), with_state_file(path), with_seq_seed(9))
            p2, s2 = hlc2.now()

        self.assertTrue(p2 > p1 or (p2 == p1 and s2 > s1))

    def test_shared_state_file(self) -> None:
        current = {"value": 1000}
        with tempfile.TemporaryDirectory() as tmp:
            path = os.path.join(tmp, "state.bin")
            # 多写者活跃共享：必须显式开启严格模式（每次发号读盘取 max）
            hlc1 = HLC(with_clock(lambda: current["value"]), with_state_file(path), with_seq_seed(9), with_state_strict())
            hlc2 = HLC(with_clock(lambda: current["value"]), with_state_file(path), with_seq_seed(9), with_state_strict())

            prev_p, prev_s = hlc1.now()
            for i in range(1000):
                if i % 2 == 0:
                    p, s = hlc2.now()
                else:
                    p, s = hlc1.now()
                self.assertTrue(p > prev_p or (p == prev_p and s > prev_s))
                prev_p, prev_s = p, s

    def test_corrupted_tail(self) -> None:
        current = {"value": 1000}
        with tempfile.TemporaryDirectory() as tmp:
            path = os.path.join(tmp, "state.bin")
            hlc = HLC(with_clock(lambda: current["value"]), with_state_file(path))
            hlc.now()
            hlc.close()  # 快速路径为批量缓冲：先落盘有效记录

            with open(path, "ab") as f:
                f.write(b"ABCDEFGHIJKLMNOPQR")  # 18 字节 CRC 不匹配的坏记录

            hlc2 = HLC(with_clock(lambda: current["value"]), with_state_file(path))
            hlc2.now()  # 坏损尾部被跳过，继续工作

    def test_node_id_out_of_range(self) -> None:
        hlc = HLC(with_node_count(3))  # workerBits=2，nodeID 上限 3
        with self.assertRaises(ValueError):
            Generator(4, hlc)

    def test_generator_monotonic(self) -> None:
        hlc = HLC()
        gen = Generator(1, hlc)
        prev = gen.next()
        for _ in range(1000):
            cur = gen.next()
            self.assertGreater(cur, prev)
            prev = cur

    def test_generator_concurrent_unique(self) -> None:
        hlc = HLC()
        gen = Generator(1, hlc)
        n = 10000
        ids = []
        lock = threading.Lock()

        def worker() -> None:
            for _ in range(n // 8):
                value = gen.next()
                with lock:
                    ids.append(value)

        threads = [threading.Thread(target=worker) for _ in range(8)]
        for t in threads:
            t.start()
        for t in threads:
            t.join()

        self.assertEqual(len(ids), n)
        self.assertEqual(len(set(ids)), n)

    def test_generator_fields(self) -> None:
        current = {"value": 1767225600123}
        hlc = HLC(with_clock(lambda: current["value"]), with_seq_seed(9))
        gen = Generator(7, hlc)

        value = gen.next()
        identifier = ID.from_int(value)
        self.assertEqual(identifier.node_id(gen.worker_bits), 7)


if __name__ == "__main__":
    unittest.main()

File: quant1x/id/id64/id_test.go
package id64

import (
	"os"
	"path/filepath"
	"sync"
	"testing"
)

func TestHLC_RollbackMonotonic(t *testing.T) {
	now := int64(1000)
	hlc := NewHLC(WithClock(func() int64 { return now }), WithSeqSeed(9))
	p1, s1 := hlc.Now()

	now = 500 // 时钟回拨
	p2, s2 := hlc.Now()
	if p2 < p1 || (p2 == p1 && s2 <= s1) {
		t.Fatalf("回拨后未保持单调: (%d,%d) -> (%d,%d)", p1, s1, p2, s2)
	}
}

func TestHLC_UsesOptionsAtConstruction(t *testing.T) {
	now := int64(4321)
	hlc := NewHLC(WithClock(func() int64 { return now }), WithSeqSeed(9))
	if hlc.Timestamp() != 4321 {
		t.Fatalf("Timestamp() = %d, want 4321", hlc.Timestamp())
	}
	if hlc.seq != 9 {
		t.Fatalf("initial seq = %d, want 9", hlc.seq)
	}
}

func TestHLC_NodeCountDerivation(t *testing.T) {
	cases := []struct {
		count      uint32
		workerBits uint8
		seqBits    uint8
	}{
		{1024, 11, 11},
		{5000, 13, 9},
		{3, 2, 20},
		{131072, 18, 4},
	}
	for _, tc := range cases {
		hlc := NewHLC(WithNodeCount(tc.count))
		if hlc.SeqBits() != tc.seqBits {
			t.Fatalf("count=%d: SeqBits()=%d, want %d", tc.count, hlc.SeqBits(), tc.seqBits)
		}
		gen := NewGenerator(0, hlc)
		if gen.WorkerBits() != tc.workerBits {
			t.Fatalf("count=%d: WorkerBits()=%d, want %d", tc.count, gen.WorkerBits(), tc.workerBits)
		}
	}
}

func TestHLC_NodeCountTooLarge(t *testing.T) {
	defer func() {
		if r := recover(); r == nil {
			t.Fatal("节点数超限应 panic")
		}
	}()
	NewHLC(WithNodeCount(262144)) // seqBits = 3 < 4
}

func TestID_FieldDecoding(t *testing.T) {
	elapsed := int64(0x123456789A)
	workerBits := uint8(11)
	seqBits := uint8(11)
	nodeID := uint32(0x1F)
	seq := uint32(0x2A)

	id := ID(uint64(elapsed)<<payloadBits | uint64(nodeID)<<seqBits | uint64(seq))
	if got := id.Physical(); got != elapsed {
		t.Fatalf("Physical() = %d, want %d", got, elapsed)
	}
	if got := id.NodeID(workerBits); got != nodeID {
		t.Fatalf("NodeID() = %d, want %d", got, nodeID)
	}
	if got := id.Seq(workerBits); got != seq {
		t.Fatalf("Seq() = %d, want %d", got, seq)
	}
	if b := id.Bytes(); FromBytes(b) != id {
		t.Fatalf("Bytes/FromBytes 往返不一致: %d", id)
	}
}

func TestID_StringRoundTrip(t *testing.T) {
	hlc := NewHLC(WithClock(func() int64 { return 1767225600123 }))
	gen := NewGenerator(7, hlc)
	id := gen.Next()
	if id.String() == "" {
		t.Fatal("String() 为空")
	}
}

func TestHLC_PersistentStateAcrossRestart(t *testing.T) {
	now := int64(1000)
	path := filepath.Join(t.TempDir(), "state.bin")
	hlc1 := NewHLC(WithClock(func() int64 { return now }), WithStateFile(path), WithSeqSeed(9))
	p1, s1 := hlc1.Now()
	// 快速路径为批量缓冲：优雅退出前 Close 刷盘，确保重启恢复最新水位
	if err := hlc1.Close(); err != nil {
		t.Fatal(err)
	}

	hlc2 := NewHLC(WithClock(func() int64 { return now }), WithStateFile(path), WithSeqSeed(9))
	p2, s2 := hlc2.Now()
	if p2 < p1 || (p2 == p1 && s2 <= s1) {
		t.Fatalf("重启后未保持单调: (%d,%d) -> (%d,%d)", p1, s1, p2, s2)
	}
}

func TestHLC_SharedStateFile(t *testing.T) {
	now := int64(1000)
	path := filepath.Join(t.TempDir(), "state.bin")
	// 多写者活跃共享：必须显式开启严格模式（每次发号读盘取 max）
	hlc1 := NewHLC(WithClock(func() int64 { return now }), WithStateFile(path), WithSeqSeed(9), WithStateStrict())
	defer hlc1.Close()
	hlc2 := NewHLC(WithClock(func() int64 { return now }), WithStateFile(path), WithSeqSeed(9), WithStateStrict())
	defer hlc2.Close()

	prevP, prevS := hlc1.Now()
	for i := 0; i < 1000; i++ {
		var p int64
		var s uint32
		if i%2 == 0 {
			p, s = hlc2.Now()
		} else {
			p, s = hlc1.Now()
		}
		if p < prevP || (p == prevP && s <= prevS) {
			t.Fatalf("共享状态文件下未保持单调: (%d,%d) -> (%d,%d)", prevP, prevS, p, s)
		}
		prevP, prevS = p, s
	}
}

func TestHLC_CorruptedTailTruncation(t *testing.T) {
	now := int64(1000)
	path := filepath.Join(t.TempDir(), "state.bin")
	hlc := NewHLC(WithClock(func() int64 { return now }), WithStateFile(path))
	hlc.Now()
	// 快速路径为批量缓冲：先 Close 落盘，保证文件存在且含有效记录
	if err := hlc.Close(); err != nil {
		t.Fatal(err)
	}

	f, err := os.OpenFile(path, os.O_APPEND|os.O_WRONLY, 0o644)
	if err != nil {
		t.Fatal(err)
	}
	// 18 字节、CRC 不匹配的坏记录，触发截断
	if _, err := f.Write([]byte("ABCDEFGHIJKLMNOPQR")); err != nil {
		t.Fatal(err)
	}
	f.Close()

	hlc2 := NewHLC(WithClock(func() int64 { return now }), WithStateFile(path))
	hlc2.Now() // 坏损尾部应被截断并继续工作（不 panic）
}

func TestGenerator_NodeIDOutOfRange(t *testing.T) {
	hlc := NewHLC(WithNodeCount(3)) // workerBits=2，nodeID 上限 3
	defer func() {
		if r := recover(); r == nil {
			t.Fatal("nodeID 越界应 panic")
		}
	}()
	NewGenerator(4, hlc)
}

func TestGenerator_Monotonic(t *testing.T) {
	hlc := NewHLC()
	gen := NewGenerator(1, hlc)
	prev := gen.Next()
	for i := 0; i < 1000; i++ {
		next := gen.Next()
		if next <= prev {
			t.Fatalf("ID 未保持单调: %d -> %d", prev, next)
		}
		prev = next
	}
}

func TestGenerator_ConcurrentUnique(t *testing.T) {
	hlc := NewHLC()
	gen := NewGenerator(1, hlc)
	const n = 10000
	ids := make(chan ID, n)
	var wg sync.WaitGroup
	for i := 0; i < 8; i++ {
		wg.Add(1)
		go func() {
			defer wg.Done()
			for j := 0; j < n/8; j++ {
				ids <- gen.Next()
			}
		}()
	}
	wg.Wait()
	close(ids)

	seen := make(map[ID]bool, n)
	for id := range ids {
		if seen[id] {
			t.Fatalf("重复 ID: %d", id)
		}
		seen[id] = true
	}
}

func BenchmarkNext(b *testing.B) {
	hlc := NewHLC()
	gen := NewGenerator(1, hlc)
	b.ReportAllocs()
	for i := 0; i < b.N; i++ {
		_ = gen.Next()
	}
}

func BenchmarkNextStateFile(b *testing.B) {
	hlc := NewHLC(WithStateFile(filepath.Join(b.TempDir(), "state.bin")))
	gen := NewGenerator(1, hlc)
	b.ReportAllocs()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_ = gen.Next()
	}
}

// 严格模式：每次发号读盘取 max（多写者活跃共享唯一性）
func BenchmarkNextStateFileStrict(b *testing.B) {
	hlc := NewHLC(WithStateFile(filepath.Join(b.TempDir(), "state.bin")), WithStateStrict())
	defer hlc.Close()
	gen := NewGenerator(1, hlc)
	b.ReportAllocs()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		_ = gen.Next()
	}
}

// 快速路径批量缓冲语义：未攒满不落盘、攒满自动落盘、Close 刷盘后可重启恢复。
func TestHLC_BatchBufferFlushSemantics(t *testing.T) {
	now := int64(1000)
	path := filepath.Join(t.TempDir(), "state.bin")
	hlc := NewHLC(
		WithClock(func() int64 { return now }),
		WithStateFile(path),
		WithStateSyncEvery(10),
	)

	// 未攒满：状态文件不应落盘
	for i := 0; i < 5; i++ {
		hlc.Now()
	}
	if _, err := os.Stat(path); !os.IsNotExist(err) {
		t.Fatalf("state file should not exist before batch flush, stat err = %v", err)
	}

	// 攒满 syncEvery：自动落盘一次
	for i := 0; i < 5; i++ {
		hlc.Now()
	}
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("state file should exist after batch flush: %v", err)
	}
	if len(data) < 10*persistentStateRecordSize {
		t.Fatalf("state file should contain >= 10 records, got %d bytes", len(data))
	}

	// Close 刷盘剩余缓冲后，重启恢复最新水位
	if err := hlc.Close(); err != nil {
		t.Fatalf("Close() error = %v", err)
	}
	prevP, prevS := hlc.Now()
	if err := hlc.Close(); err != nil {
		t.Fatalf("Close() (idempotent) error = %v", err)
	}
	hlc2 := NewHLC(
		WithClock(func() int64 { return now }),
		WithStateFile(path),
		WithStateSyncEvery(10),
	)
	p2, s2 := hlc2.Now()
	if p2 < prevP || (p2 == prevP && s2 <= prevS) {
		t.Fatalf("restart state did not advance: prev=(%d,%d) next=(%d,%d)", prevP, prevS, p2, s2)
	}
}
